library/modules/Job.cpp

```diff
@@ -410,6 +422,24 @@ bool DFHack::Job::removeJob(df::job* job) {
     return true;
 }
 
+bool DFHack::Job::addWorker(df::job *job, df::unit *unit){
```

Comment:
Yes, I had actually seen this discrepancy in `removePostings`, I was just not sure whether I should act on this or  not. I will act later today. On the topic of `removedPostings`, my understanding is that there should never be multiple postings for one job, am not entirely sure what to think of the second half of this function. 

> in general there should never exist a jobst object that is not properly linked onto the global job list. if the job's list_link member is not properly initialized, the destructor will have UB. our constructor should probably mirror this behavior if it does not already

I had wanted to add a `createJob` factory for this behavior. However, in light of [this discussion](https://discord.com/channels/793331351645323264/807443262704058408/1254199886757367808), I refrained from doing so. So now I have received opinions ranging from "there shouldn't even be a factory" to "this should be the behavior of the constructor" @lethosor , could you please comment on this?

Reply:
I did just check, and the game's `job` destructor will _not_ attempt to dereference the `job_list_link` pointer if it is null, so it's not absolutely essential to link the job; at the same time I believe our constructor should mirror DF's constructor.

The game won't malfunction if a freshly-constructed job is left on the job list; it'll have a job type of -1 (`NONE`) which the game will ignore.

Reply:
The issue here seems to be a question of sequencing. You seem to want to insist that a job be "fully initialized" before it is put on the global job list, while DF itself puts the job on the global job list when it's initially constructed, before the rest of its values are initialized.

The problem here is that when a job is being created in order to immediately assign it to a specific unit, your opinion mandates assigning the job to the unit before the job is put on the global job list. This is the opposite of the sequence DF itself uses (it creates the job, puts it on the global job list, and then assigns it to the unit). 

I think the proper sequence should be "create job, link job to global job list, link job to unit", which is what this code implements. If for some reason the last step is not done, the game state will remain safe. You seem to be saying that the sequence should be "create job, link job to unit, link job to global job list", which I disagree with because if for some reason the last step is not done, the game state may be corrupted. I am confident the game will not malfunction if a "blank" job is accidentally on the global job list, I have no faith at this time that it will not malfunction if a job has been assigned to a unit that is not also properly ensconced on the global job list. Specifically, I suspect that if the game is saved while in this state, the saved unit will be serialized with a current job ID that will not exist in the save, which may cause the game to crash when that save is subsequently loaded. I am thus concerned that the position you're taking on this issue (which I believe you think is intended to reduce risk) actually _increases_ the risk of causing UB.

Reply:
@lethosor 
> My impression is that this PR isn't touching the job constructor at all, and I don't think its scope necessarily needs to expand (i.e. the question of what our job construction logic should do shouldn't block this PR).

That is correct, and indeed I would be happy if this could be merged. However, the reason this doesn't touch job creation is that I was aware of your concerns when creating this PR. This is why I raised the issue.